04-using-github-with-the-platform.Rmd

```diff
@@ -192,7 +193,89 @@ The platform has configured simple "safety barriers" to reduce risk of accidenta
+
+As a result, sometime Private repos can be necessary, for example when it reveals a sensitive policy change that is not yet announced.
+
+If in doubt, discuss with your manager and/or the AP team.
+
+### Private R packages for reproducible analysis
+
+When a repo (e.g. an R package) is private you need to use a Personal Access Token (PAT) to access it from R. This token acts much like a password to your personal GitHub account. If you don't use the token then you get: `Error: HTTP error 404. Not Found`. (GitHub doesn't even acknowledge the existence of the repo, to avoid speculative searching for private repositories.)
+
+### Generating a PAT
+
+You can get a PAT from your GitHub user profile. A user can generate a number of tokens that grant various different levels of permission.
+
+**The PAT must be kept secret and protected like a password. It must not be shared with other users. Only generate your PAT to have the minimum permissions needed for the job.**
+
+To generate your PAT for reading private repos in R:
+
+  1. Navigate to **_Settings > Developer Settings > Personal Access Tokens_**
+  !['PAT' page on GitHub](images/pat/pat-settings.png)
+  
+  2. Select **_Generate new token_** and select the **scope** of access you want to grant using this token.
+  
+     **IMPORTANT: Only select the first group: 'repo'. Granting other rights can be dangerous if your PAT falls into the wrong hands, allowing someone else to irrevocably delete your repos, read and change your user profile settings, or even access billing information. If you need to to use other rights, consult with the AP team for advice on security arrangements to protect the PAT.**
+  !['New PAT' page](images/pat/pat-scope.png)
+  
+  3. Copy your PAT to the clipboard. It is now ready to paste directly into the place it will be used, which is likely to be your R or Python environment - see the next section.
+  
+     Note: Although GitHub won't show this PAT again, DON'T save it elsewhere. If you need a PAT in the future (eg you wiped your .rstudio folder) simply create a new PAT, and revoke the previous one. Storing secrets unnecessarily, particularly in plain text, is a security hazard.
+  ![The page displaying the new PAT](images/pat/pat-copy.png)
+
+### Using a PAT to authenticate in R/RStudio
+
+You should store your PAT in an R profile file in your *home directory* (on AP). This file gets run when you start R, putting the PAT into the system environment variable `GITHUB_PAT`. This is where common R packages (e.g. remotes and devtools) will look for it.
+
+Set this up by running this in your R console and substitute your own PAT Token:
+  
+```{r, eval = FALSE}
+writeLines('Sys.setenv("GITHUB_PAT" = "ax451...8838b1")', con = "~/.Rprofile")
+```
+
+**Warning: This line should not be put in any repo. And this .Rprofile file should not be added to any repo.**
```

Comment:
Can we put this warning immediately before the line in question, rather than (or possibly as well as) immediately after? I worry about people greedily copying and pasting before they read on.

Reply:
Yep, in.